In [1]:
# Import the modules
import numpy as np
import pandas as pd
from pathlib import Path
from sklearn.metrics import balanced_accuracy_score, confusion_matrix, classification_report

---

## Split the Data into Training and Testing Sets

### Step 1: Read the `lending_data.csv` data from the `Resources` folder into a Pandas DataFrame.

In [2]:
# Read the CSV file from the Resources folder into a Pandas DataFrame
file_path = Path("Resources/lending_data.csv")  
data = pd.read_csv(file_path)

# Review the DataFrame
data.head()  # Display the first few rows of the DataFrame


loan_size  interest_rate  borrower_income  debt_to_income  num_of_accounts  \
0    10700.0          7.672            52800        0.431818                5   
1     8400.0          6.692            43600        0.311927                3   
2     9000.0          6.963            46100        0.349241                3   
3    10700.0          7.664            52700        0.430740                5   
4    10800.0          7.698            53000        0.433962                5   

   derogatory_marks  total_debt  loan_status  
0                 1       22800            0  
1                 0       13600            0  
2                 0       16100            0  
3                 1       22700            0  
4                 1       23000            0

### Step 2: Create the labels set (`y`)  from the “loan_status” column, and then create the features (`X`) DataFrame from the remaining columns.

In [3]:
# Separate the data into labels and features

# Separate the y variable, the labels
y = data["loan_status"]

# Separate the X variable, the features
X = data.drop("loan_status", axis=1)  # Remove the "loan_status" column from the DataFrame


In [4]:
# Review the y variable Series
print(y)


0        0
1        0
2        0
3        0
4        0
        ..
77531    1
77532    1
77533    1
77534    1
77535    1
Name: loan_status, Length: 77536, dtype: int64


In [5]:
# Review the X variable DataFrame
print(X)


       loan_size  interest_rate  borrower_income  debt_to_income  \
0        10700.0          7.672            52800        0.431818   
1         8400.0          6.692            43600        0.311927   
2         9000.0          6.963            46100        0.349241   
3        10700.0          7.664            52700        0.430740   
4        10800.0          7.698            53000        0.433962   
...          ...            ...              ...             ...   
77531    19100.0         11.261            86600        0.653580   
77532    17700.0         10.662            80900        0.629172   
77533    17600.0         10.595            80300        0.626401   
77534    16300.0         10.068            75300        0.601594   
77535    15600.0          9.742            72300        0.585062   

       num_of_accounts  derogatory_marks  total_debt  
0                    5                 1       22800  
1                    3                 0       13600  
2                 

### Step 3: Check the balance of the labels variable (`y`) by using the `value_counts` function.

In [6]:
# Check the balance of our target values
target_balance = y.value_counts()
print(target_balance)


0    75036
1     2500
Name: loan_status, dtype: int64


### Step 4: Split the data into training and testing datasets by using `train_test_split`.

In [7]:
# Import the train_test_split module
from sklearn.model_selection import train_test_split

# Split the data using train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=1)


---

## Create a Logistic Regression Model with the Original Data

###  Step 1: Fit a logistic regression model by using the training data (`X_train` and `y_train`).

In [8]:
# Import the LogisticRegression module from SKLearn
from sklearn.linear_model import LogisticRegression

# Instantiate the Logistic Regression model
logreg_model = LogisticRegression(random_state=1)

# Fit the model using training data
logreg_model.fit(X_train, y_train)


LogisticRegression(random_state=1)

### Step 2: Save the predictions on the testing data labels by using the testing feature data (`X_test`) and the fitted model.

In [9]:
# Make a prediction using the testing data
y_pred = logreg_model.predict(X_test)


### Step 3: Evaluate the model’s performance by doing the following:

* Calculate the accuracy score of the model.

* Generate a confusion matrix.

* Print the classification report.

In [10]:
# Print the balanced_accuracy score of the model
from sklearn.metrics import accuracy_score

accuracy = accuracy_score(y_test, y_pred)
print(f"Accuracy: {accuracy:.2f}")


Accuracy: 0.99


In [11]:
# Generate a confusion matrix for the model
from sklearn.metrics import confusion_matrix

conf_matrix = confusion_matrix(y_test, y_pred)
print("Confusion Matrix:")
print(conf_matrix)


Confusion Matrix:
[[14926    75]
 [   46   461]]


In [12]:
# Print the classification report for the model
from sklearn.metrics import classification_report

class_report = classification_report(y_test, y_pred)
print("Classification Report:")
print(class_report)


Classification Report:
              precision    recall  f1-score   support

           0       1.00      1.00      1.00     15001
           1       0.86      0.91      0.88       507

    accuracy                           0.99     15508
   macro avg       0.93      0.95      0.94     15508
weighted avg       0.99      0.99      0.99     15508



### Step 4: Answer the following question.

**Question:** How well does the logistic regression model predict both the `0` (healthy loan) and `1` (high-risk loan) labels?

**Answer:** Question: How well does the logistic regression model predict both the 0 (healthy loan) and 1 (high-risk loan) labels?

Answer: The logistic regression model's performance in predicting both the "healthy loan" (0) and "high-risk loan" (1) labels is assessed using the balanced accuracy score, confusion matrix, and classification report.

1. Balanced Accuracy Score: The balanced accuracy score of the model is 0.99. This high score indicates that the model performs well in predicting both the majority class (healthy loans) and the minority class (high-risk loans).

2. Confusion Matrix: The confusion matrix provides insight into the model's performance across different prediction outcomes. For the "healthy loan" class (0), there are 14,926 true positives and 75 false negatives. For the "high-risk loan" class (1), there are 461 true positives and 46 false negatives.

   Confusion Matrix:
   [[14926    75]
    [   46   461]]

3. Classification Report: The classification report includes precision, recall, F1-score, and support metrics for both classes. For the "healthy loan" class (0), the model achieves high precision, recall, and F1-score values, all around 1.00. For the "high-risk loan" class (1), the model shows a lower precision of 0.86, suggesting that some "high-risk loan" predictions might be misclassified as "healthy loans." The recall for the "high-risk loan" class is 0.91, indicating that the model successfully identifies a significant portion of actual high-risk loans.

   Classification Report:
                 precision    recall  f1-score   support
   0             1.00         1.00      1.00     15001
   1             0.86         0.91      0.88       507

In summary, the logistic regression model demonstrates exceptional performance in predicting the "healthy loan" (0) class with high accuracy, precision, recall, and F1-score. While its performance on the "high-risk loan" (1) class is slightly lower, with a lower precision, the model still achieves a good balance between precision and recall, correctly identifying a significant proportion of high-risk loans.


---

## Predict a Logistic Regression Model with Resampled Training Data

### Step 1: Use the `RandomOverSampler` module from the imbalanced-learn library to resample the data. Be sure to confirm that the labels have an equal number of data points. 

In [13]:
# Import the RandomOverSampler module from imbalanced-learn
from imblearn.over_sampling import RandomOverSampler

# Instantiate the random oversampler model
random_oversampler = RandomOverSampler(random_state=1)

# Fit the original training data to the random_oversampler model
X_resampled, y_resampled = random_oversampler.fit_resample(X_train, y_train)


In [14]:
# Count the distinct values of the resampled labels data
resampled_label_counts = y_resampled.value_counts()
print(resampled_label_counts)


0    60035
1    60035
Name: loan_status, dtype: int64


### Step 2: Use the `LogisticRegression` classifier and the resampled data to fit the model and make predictions.

In [15]:
# Instantiate the Logistic Regression model
logreg_model_resampled = LogisticRegression(random_state=1)

# Fit the model using the resampled training data
logreg_model_resampled.fit(X_resampled, y_resampled)

# Make a prediction using the testing data
y_pred_resampled = logreg_model_resampled.predict(X_test)


### Step 3: Evaluate the model’s performance by doing the following:

* Calculate the accuracy score of the model.

* Generate a confusion matrix.

* Print the classification report.

In [16]:
# Print the balanced_accuracy score of the model 
from sklearn.metrics import accuracy_score

accuracy_resampled = accuracy_score(y_test, y_pred_resampled)
print(f"Accuracy (Resampled Data): {accuracy_resampled:.2f}")


Accuracy (Resampled Data): 0.99


In [17]:
# Generate a confusion matrix for the model
from sklearn.metrics import confusion_matrix

conf_matrix_resampled = confusion_matrix(y_test, y_pred_resampled)
print("Confusion Matrix (Resampled Data):")
print(conf_matrix_resampled)


Confusion Matrix (Resampled Data):
[[14915    86]
 [    3   504]]


In [18]:
# Print the classification report for the model
from sklearn.metrics import classification_report

class_report_resampled = classification_report(y_test, y_pred_resampled)
print("Classification Report (Resampled Data):")
print(class_report_resampled)


Classification Report (Resampled Data):
              precision    recall  f1-score   support

           0       1.00      0.99      1.00     15001
           1       0.85      0.99      0.92       507

    accuracy                           0.99     15508
   macro avg       0.93      0.99      0.96     15508
weighted avg       1.00      0.99      0.99     15508



### Step 4: Answer the following question

**Question:** How well does the logistic regression model, fit with oversampled data, predict both the `0` (healthy loan) and `1` (high-risk loan) labels?

**Answer:** Question: How well does the logistic regression model, fit with oversampled data, predict both the 0 (healthy loan) and 1 (high-risk loan) labels?

Answer: The logistic regression model, fit with oversampled data, demonstrates exceptional performance in predicting both the "healthy loan" (0) and "high-risk loan" (1) labels. This is evident from the evaluation metrics obtained after training the model with resampled data.

1. Balanced Accuracy Score: The balanced accuracy score of the model trained with oversampled data is 0.99. This high score indicates the model's remarkable ability to correctly predict both majority and minority classes, showcasing its effectiveness in handling imbalanced data.

2. Confusion Matrix: The confusion matrix illustrates the distribution of true positive, true negative, false positive, and false negative predictions for each class. For the "healthy loan" class (0), there are 14,915 true positives and 86 false negatives. For the "high-risk loan" class (1), there are 504 true positives and 3 false negatives.

   Confusion Matrix (Resampled Data):
   [[14915    86]
    [    3   504]]

3. Classification Report: The classification report provides precision, recall, F1-score, and support metrics for both classes. The "healthy loan" class (0) achieves a precision, recall, and F1-score of 1.00, indicating near-perfect performance. The "high-risk loan" class (1) displays a precision of 0.85 and a recall of 0.99, signifying the model's strong capability to correctly identify high-risk loans.

   Classification Report (Resampled Data):
                 precision    recall  f1-score   support
   0             1.00         0.99      1.00     15001
   1             0.85         0.99      0.92       507

In summary, the logistic regression model, fit with oversampled data, excels in predicting both "healthy loan" and "high-risk loan" labels. It achieves exceptional accuracy, precision, recall, and F1-score metrics, showcasing its robustness in handling imbalanced classes and effectively identifying high-risk loans while maintaining strong performance on healthy loans.
